File: MacApp/Sources/VTubeLinkService/iFacialMocapReceiver.swift
import Foundation
import Network
import VTubeLinkShared

class IFacialMocapReceiver: ObservableObject {
    @Published var isReceiving = false
    @Published var latestData = CapturedData()
    
    private var listener: NWListener?
    private var connection: NWConnection?
    private let port: NWEndpoint.Port = 49983
    
    func start(ipAddress: String) {
        guard !ipAddress.isEmpty else { return }
        
        // 1. Send the init ping to the iPhone
        let host = NWEndpoint.Host(ipAddress)
        connection = NWConnection(host: host, port: port, using: .udp)
        connection?.stateUpdateHandler = { state in
            switch state {
            case .ready:
                let initString = "iFacialMocap_sahuasouryya9218sauhuiayeta91555dy3719"
                let data = initString.data(using: .utf8)
                self.connection?.send(content: data, completion: .contentProcessed({ error in
                    if let error = error {
                        print("UDP send error: \(error)")
                    } else {
                        print("Sent initialization string to iFacialMocap")
                    }
                }))
            default:
                break
            }
        }
        connection?.start(queue: .global())
        
        // 2. Start listening on port 49983 for incoming data
        do {
            listener = try NWListener(using: .udp, on: port)
            listener?.stateUpdateHandler = { state in
                print("Listener state: \(state)")
            }
            listener?.newConnectionHandler = { newConnection in
                newConnection.start(queue: .global())
                self.receiveMessage(on: newConnection)
            }
            listener?.start(queue: .global())
            DispatchQueue.main.async {
                self.isReceiving = true
            }
        } catch {
            print("Failed to create listener: \(error)")
        }
    }
    
    func stop() {
        listener?.cancel()
        listener = nil
        connection?.cancel()
        connection = nil
        DispatchQueue.main.async {
            self.isReceiving = false
        }
    }
    
    private func receiveMessage(on connection: NWConnection) {
        connection.receiveMessage { data, context, isComplete, error in
            if let data = data, let message = String(data: data, encoding: .utf8) {
                self.processData(message)
                self.receiveMessage(on: connection) // Continue listening
            } else if let error = error {
                print("Receive error: \(error)")
            }
        }
    }
    
    private func processData(_ rawData: String) {
        var paramsDict: [String: [Float]] = [:]
        
        let paramStrs = rawData.trimmingCharacters(in: CharacterSet(charactersIn: "|")).components(separatedBy: "|")
        for paramStr in paramStrs {
            if paramStr.contains("#") {
                let parts = paramStr.components(separatedBy: "#")
                guard parts.count == 2 else { continue }
                let key = parts[0]
                let valStrs = parts[1].components(separatedBy: ",")
                let vals = valStrs.compactMap { Float($0) }
                paramsDict[key] = vals
            } else if paramStr.contains("-") {
                let parts = paramStr.components(separatedBy: "-")
                guard parts.count == 2 else { continue }
                let key = parts[0]
                if let val = Float(parts[1]) {
                    paramsDict[key] = [val / 100.0]
                }
            }
        }
        
        var newData = CapturedData()
        
        for blendshape in Constants.blendshapeNames {
            if let val = paramsDict[blendshape]?.first {
                newData.blendshapes[blendshape] = val
            } else {
                newData.blendshapes[blendshape] = 0.0
            }
        }
        
        if let headVals = paramsDict["=head"], headVals.count >= 6 {
            newData.headRotationX = headVals[0]
            newData.headRotationY = headVals[1]
            newData.headRotationZ = headVals[2]
            newData.headPositionX = headVals[3]
            newData.headPositionY = headVals[4]
            newData.headPositionZ = headVals[5]
        }
        
        if let rightEye = paramsDict["rightEye"], rightEye.count >= 3 {
            newData.rightEyeRotationX = rightEye[0]
            newData.rightEyeRotationY = rightEye[1]
            newData.rightEyeRotationZ = rightEye[2]
        }
        
        if let leftEye = paramsDict["leftEye"], leftEye.count >= 3 {
            newData.leftEyeRotationX = leftEye[0]
            newData.leftEyeRotationY = leftEye[1]
            newData.leftEyeRotationZ = leftEye[2]
        }
        
        DispatchQueue.main.async {
            self.latestData = newData
        }
    }
}

File: WinApp/IFacialMocapReceiver.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Globalization;

namespace VTubeLink
{
    public class IFacialMocapReceiver : INotifyPropertyChanged
    {
        private bool _isReceiving;
        public bool IsReceiving
        {
            get => _isReceiving;
            private set
            {
                if (_isReceiving != value)
                {
                    _isReceiving = value;
                    PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsReceiving)));
                }
            }
        }

        private CapturedData _latestData = new();
        public CapturedData LatestData
        {
            get => _latestData;
            private set
            {
                _latestData = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(LatestData)));
            }
        }

        public event PropertyChangedEventHandler? PropertyChanged;

        private UdpClient? _udpClient;
        private CancellationTokenSource? _cancellationTokenSource;
        private const int Port = 49983;

        public async Task StartAsync(string ipAddress)
        {
            if (string.IsNullOrWhiteSpace(ipAddress)) return;

            Stop();

            _cancellationTokenSource = new CancellationTokenSource();
            var token = _cancellationTokenSource.Token;

            try
            {
                _udpClient = new UdpClient();
                // Send init ping to iPhone
                var initString = "iFacialMocap_sahuasouryya9218sauhuiayeta91555dy3719";
                var initBytes = Encoding.UTF8.GetBytes(initString);
                _udpClient.Connect(ipAddress, Port);
                await _udpClient.SendAsync(initBytes, initBytes.Length);
                AppLogger.Log("UDP", "Sent initialization string to iFacialMocap");

                // Start listening
                _udpClient.Close(); // Close the connected client
                _udpClient = new UdpClient(Port); // Reopen for listening on the port
                IsReceiving = true;

                _ = Task.Run(async () =>
                {
                    try
                    {
                        while (!token.IsCancellationRequested)
                        {
                            var result = await _udpClient.ReceiveAsync(token);
                            var message = Encoding.UTF8.GetString(result.Buffer);
                            ProcessData(message);
                        }
                    }
                    catch (OperationCanceledException) { }
                    catch (SocketException ex)
                    {
                        AppLogger.Log("UDP", $"SocketException: {ex.Message}");
                    }
                    finally
                    {
                        IsReceiving = false;
                    }
                }, token);
            }
            catch (Exception ex)
            {
                AppLogger.Log("UDP", $"Failed to start receiver: {ex.Message}");
                IsReceiving = false;
            }
        }

        public void Stop()
        {
            _cancellationTokenSource?.Cancel();
            _cancellationTokenSource?.Dispose();
            _cancellationTokenSource = null;

            _udpClient?.Close();
            _udpClient?.Dispose();
            _udpClient = null;

            IsReceiving = false;
        }

        private void ProcessData(string rawData)
        {
            var paramsDict = new System.Collections.Generic.Dictionary<string, float[]>();

            var paramStrs = rawData.Trim('|').Split('|');
            foreach (var paramStr in paramStrs)
            {
                if (paramStr.Contains('#'))
                {
                    var parts = paramStr.Split('#');
                    if (parts.Length == 2)
                    {
                        var key = parts[0];
                        var valStrs = parts[1].Split(',');
                        var vals = new float[valStrs.Length];
                        for (int i = 0; i < valStrs.Length; i++)
                        {
                            if (float.TryParse(valStrs[i], NumberStyles.Float, CultureInfo.InvariantCulture, out float val))
                            {
                                vals[i] = val;
                            }
                        }
                        paramsDict[key] = vals;
                    }
                }
                else if (paramStr.Contains('-'))
                {
                    var parts = paramStr.Split('-');
                    if (parts.Length == 2)
                    {
                        var key = parts[0];
                        if (float.TryParse(parts[1], NumberStyles.Float, CultureInfo.InvariantCulture, out float val))
                        {
                            paramsDict[key] = [val / 100.0f];
                        }
                    }
                }
            }

            var newData = new CapturedData();

            foreach (var blendshape in Constants.BlendshapeNames)
            {
                if (paramsDict.TryGetValue(blendshape, out var vals) && vals.Length > 0)
                {
                    newData.Blendshapes[blendshape] = vals[0];
                }
                else
                {
                    newData.Blendshapes[blendshape] = 0.0f;
                }
            }

            if (paramsDict.TryGetValue("=head", out var headVals) && headVals.Length >= 6)
            {
                newData.HeadRotationX = headVals[0];
                newData.HeadRotationY = headVals[1];
                newData.HeadRotationZ = headVals[2];
                newData.HeadPositionX = headVals[3];
                newData.HeadPositionY = headVals[4];
                newData.HeadPositionZ = headVals[5];
            }

            if (paramsDict.TryGetValue("rightEye", out var rightEye) && rightEye.Length >= 3)
            {
                newData.RightEyeRotationX = rightEye[0];
                newData.RightEyeRotationY = rightEye[1];
                newData.RightEyeRotationZ = rightEye[2];
            }

            if (paramsDict.TryGetValue("leftEye", out var leftEye) && leftEye.Length >= 3)
            {
                newData.LeftEyeRotationX = leftEye[0];
                newData.LeftEyeRotationY = leftEye[1];
                newData.LeftEyeRotationZ = leftEye[2];
            }

            LatestData = newData;
        }
    }
}
